dev/AppNotifications/AppNotificationManager.cpp

```diff
@@ -55,76 +56,121 @@ namespace winrt::Microsoft::Windows::AppNotifications::implementation
-        if (!PushNotificationHelpers::IsPackagedAppScenario())
+        HRESULT hr{ S_OK };
+        std::wstring appId{};
+
+        auto logTelemetry{ wil::scope_exit([&]() {
+            AppNotificationTelemetry::RegisterByAPI(hr, appId);
+            THROW_IF_FAILED(hr);
+        }) };
+
+        try
         {
-            std::wstring notificationId{ RetrieveNotificationAppId() };
-            if (!AppModel::Identity::IsPackagedProcess())
+            winrt::guid storedComActivatorGuid{ GUID_NULL };
+            if (!PushNotificationHelpers::IsPackagedAppScenario())
             {
-                THROW_IF_FAILED(PushNotifications_RegisterFullTrustApplication(notificationId.c_str(), GUID_NULL));
+                std::wstring notificationId{ RetrieveNotificationAppId() };
+                if (!AppModel::Identity::IsPackagedProcess())
+                {
+                    THROW_IF_FAILED(PushNotifications_RegisterFullTrustApplication(notificationId.c_str(), GUID_NULL));
 
-                storedComActivatorGuid = RegisterComActivatorGuidAndAssets();
-            }
+                    storedComActivatorGuid = RegisterComActivatorGuidAndAssets();
+                }
 
-            wil::unique_cotaskmem_string processName;
-            THROW_IF_FAILED(GetCurrentProcessPath(processName));
-            auto notificationPlatform{ PushNotificationHelpers::GetNotificationPlatform() };
-            THROW_IF_FAILED(notificationPlatform->AddToastRegistrationMapping(processName.get(), notificationId.c_str()));
-        }
+                wil::unique_cotaskmem_string processName;
+                THROW_IF_FAILED(GetCurrentProcessPath(processName));
+                auto notificationPlatform{ PushNotificationHelpers::GetNotificationPlatform() };
+                THROW_IF_FAILED(notificationPlatform->AddToastRegistrationMapping(processName.get(), notificationId.c_str()));
+            }
 
-        winrt::guid registeredClsid{ GUID_NULL };
-        if (AppModel::Identity::IsPackagedProcess())
-        {
-            registeredClsid = PushNotificationHelpers::GetComRegistrationFromRegistry(expectedAppServerArgs.data());
-        }
+            winrt::guid registeredClsid{ GUID_NULL };
+            if (AppModel::Identity::IsPackagedProcess())
+            {
+                registeredClsid = PushNotificationHelpers::GetComRegistrationFromRegistry(expectedAppServerArgs.data());
+            }
 
-        // Create event handle before COM Registration otherwise if a notification arrives will lead to race condition
-        GetWaitHandleForArgs().create();
+            // Create event handle before COM Registration otherwise if a notification arrives will lead to race condition
+            GetWaitHandleForArgs().create();
 
+            {
+                auto lock{ m_lock.lock_exclusive() };
+                THROW_HR_IF_MSG(E_INVALIDARG, m_notificationComActivatorRegistration, "Already Registered for App Notifications!");
```

Comment:
This is an example of why I'm not a fan of throwing away the existing exception and creating a new one solely based on the hr in scope_exit(). You loose context, like the message in this case.

Reply:
I've fixed this. we don't have to create a new exception with the hr. Actually, creating a new expectation would lead to a crash